File: modules/expo-audio-route/ios/ExpoAudioRouteModule.swift
import ExpoModulesCore
import AVFoundation

public class ExpoAudioRouteModule: Module {
  // Reference to the system's NotificationCenter for managing audio route change notifications.
  // Uses the default shared instance to observe AVAudioSession route change events.
  private let notificationCenter: NotificationCenter = .default

  // Observer token returned by NotificationCenter when registering for audio route change notifications.
  // This token is retained to allow proper cleanup and removal of the observer when no longer needed.
  private var routeChangeObserver: NSObjectProtocol?

  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  public func definition() -> ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ExpoAudioRoute')` in JavaScript.
    Name("ExpoAudioRoute")

    // Declares an event that JavaScript can subscribe to for audio route change notifications.
    Events("onAudioRouteChange")

    // Defines a JavaScript function that always returns a Promise and whose native code
    // is by default dispatched on the different thread than the JavaScript runtime runs on.
    // This function retrieves the current audio output route (e.g., "speaker", "bluetooth",
    // "wiredHeadset") from the device's audio session. The asynchronous nature ensures that
    // querying the AVAudioSession doesn't block the JavaScript thread, maintaining smooth UI performance.
    AsyncFunction("getCurrentRouteAsync") {
      return self.currentRoute()
    }

    // Module lifecycle callback that executes when JavaScript begins listening to the onAudioRouteChange event.
    // This is called automatically when the first event listener is added on the JavaScript side.
    OnStartObserving("onAudioRouteChange") {
      self.sendEvent("onAudioRouteChange", ["route": self.currentRoute()])
      self.startObservingRouteChanges()
    }

    // Module lifecycle callback that executes when JavaScript stops listening to the onAudioRouteChange event.
    // This is called automatically when the last event listener is removed on the JavaScript side.
    OnStopObserving("onAudioRouteChange")  {
      self.stopObservingRouteChanges()
    }
  }

  // Registers an observer with NotificationCenter to monitor AVAudioSession route changes.
  //
  // When audio routing changes (e.g., headphones plugged in/out, Bluetooth connected/disconnected),
  // the observer callback is triggered and sends an event to JavaScript with the updated route.
  //
  // Implementation details:
  // - Observes AVAudioSession.routeChangeNotification
  // - Stores the observer token in routeChangeObserver for later cleanup
  //
  // The event payload includes:
  // - route: The current audio route as determined by currentRoute()
  private func startObservingRouteChanges() {
    func handleRouteChange(_: Notification) {
      self.sendEvent("onAudioRouteChange", ["route": self.currentRoute()])
    }

    self.routeChangeObserver = NotificationCenter.default.addObserver(
      forName: AVAudioSession.routeChangeNotification,
      object: AVAudioSession.sharedInstance(),
      queue: .main,
      using: handleRouteChange
    )
  }

  // Unregisters the AVAudioSession route change notification observer.
  //
  // This cleanup function:
  // - Removes the observer from NotificationCenter to prevent memory leaks
  // - Nullifies the routeChangeObserver reference to free resources
  // - Is safe to call even if no observer is currently registered
  //
  // Should be called when:
  // - JavaScript removes all event listeners for onAudioRouteChange
  // - The module is being destroyed or cleaned up
  private func stopObservingRouteChanges() {
    if (routeChangeObserver == nil) {
      return
    }
    notificationCenter.removeObserver(routeChangeObserver!)
    routeChangeObserver = nil
  }

  // Determines the current audio output route from the AVAudioSession.
  // This method queries the AVAudioSession's current route and examines the first output port
  // to determine where audio is being routed. It maps various iOS audio port types to
  // simplified route names for easier handling in JavaScript.
  //
  // Returns: A string representing the current audio route:
  // - "wiredHeadset": Audio is routed through wired headphones or headset microphone
  // - "bluetooth": Audio is routed through any Bluetooth device (A2DP, LE, or HFP)
  // - "speaker": Audio is routed through the device's built-in speaker
  // - "unknown": No output detected or unrecognized port type
  private func currentRoute() -> String {
    let session = AVAudioSession.sharedInstance()
    let outputs = session.currentRoute.outputs

    let first = outputs.first
    if (first == nil) {
      return "unknown"
    }

    switch (first?.portType) {
      case .headphones, .headsetMic:
        return "wiredHeadset"
      case .bluetoothA2DP, .bluetoothLE, .bluetoothHFP:
        return "bluetooth"
      case .builtInSpeaker:
        return "speaker"
      default:
        return "unknown"
    }
  }
}

File: modules/expo-audio-route/android/src/main/java/expo/modules/audioroute/ExpoAudioRouteModule.kt
package expo.modules.audioroute

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import android.content.Context
import android.media.AudioDeviceInfo
import android.media.AudioManager

class ExpoAudioRouteModule : Module() {
  // Reference to the Android AudioManager system service.
  // This is initialized in the OnCreate lifecycle method and used to query audio routing information.
  private var audioManager: AudioManager? = null

  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  override fun definition() = ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ExpoAudioRoute')` in JavaScript.
    Name("ExpoAudioRoute")

    // Module lifecycle callback that executes when the module is created.
    // Initializes the audioManager by retrieving the AUDIO_SERVICE from the Android system services.
    // This ensures the AudioManager is available for all audio routing queries throughout the module's lifetime.
    OnCreate {
      audioManager = appContext.reactContext?.getSystemService(Context.AUDIO_SERVICE) as AudioManager
    }

    // Exposes the current audio route to JavaScript as an async function.
    // Can be called from JavaScript using: ExpoAudioRoute.getCurrentRouteAsync()
    AsyncFunction("getCurrentRouteAsync") {
      currentRoute()
    }
  }

  /**
  * Determines the current audio output route by querying connected audio devices.
  * Devices are checked in priority order: wired headsets → bluetooth → built-in speakers.
  *
  * The priority hierarchy ensures that:
  * - Wired connections take precedence over wireless (typically user's explicit choice)
  * - Bluetooth is preferred over built-in speakers
  * - Built-in speaker/earpiece is the fallback option
  *
  * @return String identifier of the current audio route:
  * - "wiredHeadset": Wired headphones or headset connected
  * - "bluetooth": Bluetooth A2DP or SCO device connected
  * - "speaker": Using built-in speaker or earpiece
  * - "unknown": No recognizable audio output device found or AudioManager unavailable
  *
  * Implementation details:
  * - Queries all available output devices using AudioManager.GET_DEVICES_OUTPUTS
  * - Returns "unknown" if AudioManager is null (initialization failed)
  * - Uses firstOrNull with type matching to find the highest priority connected device
  */
  private fun currentRoute(): String {
    val outputs = audioManager?.getDevices(AudioManager.GET_DEVICES_OUTPUTS)

    if(outputs.isNullOrEmpty()) return "unknown"

    // Check in priority order: wired > bluetooth > speaker
    val wiredTypes = listOf(
      AudioDeviceInfo.TYPE_WIRED_HEADPHONES,
      AudioDeviceInfo.TYPE_WIRED_HEADSET,
      AudioDeviceInfo.TYPE_USB_HEADSET
    )

    val bluetoothTypes = listOf(
      AudioDeviceInfo.TYPE_BLUETOOTH_A2DP,
      AudioDeviceInfo.TYPE_BLUETOOTH_SCO
    )

    val speakerTypes = listOf(
      AudioDeviceInfo.TYPE_BUILTIN_SPEAKER,
      AudioDeviceInfo.TYPE_BUILTIN_EARPIECE
    )

    val wired = outputs.firstOrNull { it.type in wiredTypes }
    val bluetooth = outputs.firstOrNull { it.type in bluetoothTypes }
    val speaker = outputs.firstOrNull { it.type in speakerTypes }

    val device = when {
      wired != null -> wired
      bluetooth != null -> bluetooth
      speaker != null -> speaker
      else -> null
    }

    return when (device?.type) {
      in wiredTypes -> "wiredHeadset"
      in bluetoothTypes -> "bluetooth"
      in speakerTypes -> "speaker"
      else -> "unknown"
    }
  }
}